scylla/src/transport/connection.rs

```diff
@@ -89,12 +100,38 @@ impl Connection {
         };
         let source_port = stream.local_addr()?.port();
         stream.set_nodelay(config.tcp_nodelay)?;
-
+        
         // TODO: What should be the size of the channel?
         let (sender, receiver) = mpsc::channel(128);
-
-        let (fut, _worker_handle) = Self::router(stream, receiver).remote_handle();
-        tokio::task::spawn(fut);
+        
+        let custom_stream = match config.use_tls {
+            true => {
+                let mut tls_config = ClientConfig::new();
+                tls_config
+                    .root_store
+                    .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
+                let connector = TlsConnector::from(Arc::new(tls_config));
+                let domain = addr.to_string();
+                let domain = DNSNameRef::try_from_ascii_str(&domain)
+                    .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid dnsname"))?;
+                CustomStream::Tls(connector.connect(domain, stream).await?)
+            },
+            false => CustomStream::Tcp(stream),
+        };
+        
+        let _worker_handle = match custom_stream {
+            // router returns diffrent future types so task has to be spawned separately
+            CustomStream::Tcp(stream) => {
+                let (task, handle) = Self::router(stream, receiver).remote_handle();
+                tokio::task::spawn(task);
+                handle
+            },
+            CustomStream::Tls(stream) => {
+                let (task, handle) = Self::router(stream, receiver).remote_handle();
+                tokio::task::spawn(task);
+                handle
+            },
+        };
```

Comment:
Unless you are going to use the CustomStream enum somewhere else, I think we can get rid of it. You could add a following method to the connection:

```rust

fn run_router(
    stream: (impl AsyncRead + AsyncWrite + Send + 'static),
    receiver: mpsc::Receiver<Task>,
) -> RemoteHandle<()> {
    let (task, handle) = Self::router(stream, receiver).remote_handle();
    tokio::task::spawn(task);
    handle
}
```

And change the above code to:

```rust
let _worker_handle = match config.use_tls {
    true => {
        let mut tls_config = ClientConfig::new();
        tls_config
            .root_store
            .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
        let connector = TlsConnector::from(Arc::new(tls_config));
        let domain = addr.to_string();
        let domain = DNSNameRef::try_from_ascii_str(&domain).map_err(|_| {
            std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid dnsname")
        })?;
        Self::run_router(connector.connect(domain, stream).await?, receiver)
    }
    false => Self::run_router(stream, receiver),
};
```

This gets rid of the Clippy's complaint about the enum variant sizes varying in size too much.